### PR TITLE
Support generic fields in PersistedAssemblyBuilder

### LINKED ISSUE #110247: ``PersistedAssemblyBuilder`` generates invalid IL when loading fields with generic type parameters
### Description

I have found a possible issue with the new ``PersistedAssemblyBuilder`` introduced in .NET 9 that causes invalid IL to be generated when loading fields of generic types that have a generic type parameter as the field type.

When decompiling the generated IL using ``ildasm`` or similar tools, the field reference shows up as having the concrete type instead of the type parameter.

### Reproduction Steps

The following code reproduces the problem in .NET 9. It doesn't produce a working executable by itself since I did not set the entry point, so to execute code from the generated assembly, you would need to reference it by some other assembly.
````csharp
using System;
using System.Reflection;
using System.Reflection.Emit;

PersistedAssemblyBuilder ab = new(new("a"), typeof(object).Assembly);
ModuleBuilder mb = ab.DefineDynamicModule("a");
TypeBuilder tb = mb.DefineType("Program");
MethodBuilder methb = tb.DefineMethod("Main", MethodAttributes.Public | MethodAttributes.Static, CallingConventions.Standard, typeof(void), []);

ILGenerator il = methb.GetILGenerator();
il.Emit(OpCodes.Newobj, typeof(C<int>).GetConstructor([]));
il.Emit(OpCodes.Ldfld, typeof(C<int>).GetField("F"));
il.Emit(OpCodes.Call, typeof(Console).GetMethod("WriteLine", BindingFlags.Public | BindingFlags.Static, null, [typeof(int)], []));
il.Emit(OpCodes.Ret);

tb.CreateType();
ab.Save("a.dll");

public class C<T>
{
    public C() { }
    public T F;
}
````

### Expected behavior

The following IL should be emitted:
````cil
newobj instance void class [Test]C`1<int32>::.ctor()
ldfld !0 class [Test]C`1<int32>::F
call void [mscorlib]System.Console::WriteLine(int32)
ret
````
(Above is IL generated when running the code example on .NET 4.8.1, which is why it is using ``mscorlib``.)
When the code is executed, the following output should be printed:
````
0
````

### Actual behavior

The following IL is generated:
````cil
newobj instance void class [Test]C`1<int32>::.ctor()
ldfld int32 class [Test]C`1<int32>::F
call void [System.Console]System.Console::WriteLine(int32)
ret
````
As you can see, the second line references the type ``int32`` directly, instead of the type parameter ``!0``. This causes the following exception to be thrown when this code is exectued:
````
Unhandled exception. System.MissingFieldException: Field not found: 'C`1.F'.
   at Program.Main()
````

### Regression?

API was newly introduced in .NET 9. On .NET 4.8.1, when using the regular ``AssemblyBuilder``, it generates a working assembly.

### Known Workarounds

_No response_

### Configuration

- .NET 9.0.100
- Windows 11 23H2
- Debug, x64

### Other information

I asked a question on the topic on [StackOverflow](https://stackoverflow.com/questions/79231619/how-do-i-get-a-field-definition-with-an-open-generic-return-type-from-a-closed-g), a user provided an [explanation](https://stackoverflow.com/a/79232015/13519002) which goes into more depth than I would have been able to.

### PR DESCRIPTION
**Pull request by steveharter**
Fixes https://github.com/dotnet/runtime/issues/110247

cc @jgh07

We will likely port to v9.x once verified in v10 since this was new functionality in v9.

### CODE REVIEW

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs [hunk 1]
```diff
@@ -770,6 +772,26 @@ private EntityHandle GetMemberReferenceHandle(MemberInfo memberInfo)
             }
 
             return memberHandle;
+
+            // The field type has to be the original open generic field type, not the closed type.
+            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2075:UnrecognizedReflectionPattern", Justification = "Internal reflection implementation")]
+            static Type GetFieldType(FieldInfo fieldInfo)
+            {
+                if (fieldInfo.DeclaringType!.IsGenericType)
```
Comment (buyaa-n) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs:780]:
```suggestion
                if (fieldInfo.DeclaringType!.IsConstructedGenericType)
```

**Review by buyaa-n — APPROVED** [review]
LGTM, thanks!

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs [hunk 2]
```diff
@@ -734,13 +734,15 @@ private EntityHandle GetMemberReferenceHandle(MemberInfo memberInfo)
...
-                        if (field.DeclaringType!.IsGenericTypeDefinition)
+                        if (declaringType.IsGenericTypeDefinition)
                         {
                             //The type of the field has to be fully instantiated type.
                             declaringType = declaringType.MakeGenericType(declaringType.GetGenericArguments());
                         }
+
                         memberHandle = AddMemberReference(field.Name, GetTypeHandle(declaringType),
-                            MetadataSignatureHelper.GetFieldSignature(field.FieldType, field.GetRequiredCustomModifiers(), field.GetOptionalCustomModifiers(), this));
+                            MetadataSignatureHelper.GetFieldSignature(GetFieldType(field), field.GetRequiredCustomModifiers(), field.GetOptionalCustomModifiers(), this));
```
Comment (jkotas) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs:744]: Can this use `GetOriginalMemberIfConstructedType` like the method cases? `GetOriginalMemberIfConstructedType` seems to be doing the same thing as the newly added method.

**Review by jkotas — CHANGES_REQUESTED** [review]
https://github.com/dotnet/runtime/pull/110372/files#r1876967406

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs (hunk 2 above)
Comment (buyaa-n) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs:744]: Yes `GetOriginalMemberIfConstructedType` can be used, but need to change the parameter type from `MethodBase` to `MemberInfo`

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs (hunk 1 above)
Comment (steveharter) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs]: With the switch to using GetOriginalMemberIfConstructedType(), this is no longer needed.

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs (hunk 2 above)
Comment (steveharter) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs]: Thanks - done.

**Review by jkotas — APPROVED** [review]

**Comment by ericstj** [comment]
@jkotas @steveharter - do you think we need to add more test cases to cover other permutations of generic usage?  Methods / properties / events / parameters, nested generics, etc.

**Comment by jkotas** [comment]
It is a lot of potentially interesting permutations. The chances of more broken corner-cases being found by manual test authoring are low. I think that the best way to find more broken corner-cases is roundtripping of the existing test asserts through ref.emit and validating that the roundtripped tests still pass. Buya has done some of it, but it was not completed and automated.

**Comment by buyaa-n** [comment]
> do you think we need to add more test cases to cover other permutations of generic usage? Methods / properties / events / parameters, nested generics, etc.

I added as much test as I can that covers many permutations of generic usage and nested generics for `methods and its parameter`.  Agree with Jan that to find more corner cases round tripping tool would have been helpful. As mentioned the tool was used locally, not added into the repo and now that project is gone with my old laptop. Someone could rewrite it using the .NET framework tool.

We don't have tests that calling generic `properties / events` in IL, because by my understanding `properties / events` are not called directly in IL, at least not with ref emit APIs, instead their get/set/add/remove methods are called/used in IL, and those are covered by methods tests.

**Pull request merged**